File: dqops/src/main/java/com/dqops/metadata/definitions/sensors/ProviderSensorDefinitionSpec.java
/*
 * Copyright © 2021 DQOps (support@dqops.com)
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
package com.dqops.metadata.definitions.sensors;

import com.dqops.execution.sqltemplates.rendering.JinjaSqlTemplateSensorRunner;
import com.dqops.metadata.basespecs.AbstractSpec;
import com.dqops.metadata.id.ChildHierarchyNodeFieldMap;
import com.dqops.metadata.id.ChildHierarchyNodeFieldMapImpl;
import com.dqops.metadata.id.HierarchyNodeResultVisitor;
import com.dqops.utils.serialization.InvalidYamlStatusHolder;
import com.fasterxml.jackson.annotation.JsonIgnore;
import com.fasterxml.jackson.annotation.JsonInclude;
import com.fasterxml.jackson.annotation.JsonPropertyDescription;
import com.fasterxml.jackson.databind.PropertyNamingStrategies;
import com.fasterxml.jackson.databind.annotation.JsonNaming;
import lombok.EqualsAndHashCode;

import java.util.Collections;
import java.util.Map;
import java.util.Objects;

/**
 * Specification (configuration) for a provider specific implementation of a data quality sensor or an SQL template.
 */
@JsonInclude(JsonInclude.Include.NON_NULL)
@JsonNaming(PropertyNamingStrategies.SnakeCaseStrategy.class)
@EqualsAndHashCode(callSuper = true)
public class ProviderSensorDefinitionSpec extends AbstractSpec implements InvalidYamlStatusHolder {
    private static final ChildHierarchyNodeFieldMapImpl<ProviderSensorDefinitionSpec> FIELDS = new ChildHierarchyNodeFieldMapImpl<>(AbstractSpec.FIELDS) {
        {
        }
    };

    @JsonPropertyDescription("Sensor implementation type")
    private ProviderSensorRunnerType type = ProviderSensorRunnerType.sql_template;

    @JsonPropertyDescription("Java class name for a sensor runner that will execute the sensor. The \"type\" must be \"java_class\".")
    private String javaClassName = JinjaSqlTemplateSensorRunner.CLASS_NAME;

    @JsonInclude(JsonInclude.Include.NON_NULL)
    @JsonPropertyDescription("The sensor supports grouping, using the GROUP BY clause in SQL. Sensors that support a GROUP BY condition can capture separate data quality scores for each data group. The default value is true, because most of the data quality sensor support grouping.")
    private Boolean supportsGrouping;

    @JsonInclude(JsonInclude.Include.NON_NULL)
    @JsonPropertyDescription("The sensor supports grouping by a partition date, using the GROUP BY clause in SQL. Sensors that support grouping by a partition_by_column could be used for partition checks, calculating separate data quality metrics for each daily/monthly partition. The default value is true, because most of the data quality sensor support partitioned checks.")
    private Boolean supportsPartitionedChecks;

    @JsonPropertyDescription("Additional provider specific sensor parameters")
    @JsonInclude(JsonInclude.Include.NON_EMPTY)
    private Map<String, String> parameters;

    @JsonInclude(JsonInclude.Include.NON_DEFAULT)
    @JsonPropertyDescription("Disables merging this sensor's SQL with other sensors. When this parameter is 'true', the sensor's SQL will be executed as an independent query.")
    private boolean disableMergingQueries;

    @JsonIgnore
    private String yamlParsingError;

    /**
     * Sets a value that indicates that the YAML file deserialized into this object has a parsing error.
     *
     * @param yamlParsingError YAML parsing error.
     */
    @Override
    public void setYamlParsingError(String yamlParsingError) {
        this.yamlParsingError = yamlParsingError;
    }

    /**
     * Returns the YAML parsing error that was captured.
     *
     * @return YAML parsing error.
     */
    @Override
    public String getYamlParsingError() {
        return this.yamlParsingError;
    }

    /**
     * Returns a key/value map of additional rule parameters.
     * @return Key/value dictionary of additional parameters passed to the rule.
     */
    public Map<String, String> getParameters() {
        return parameters;
    }

    /**
     * Sets a dictionary of parameters passed to the rule.
     * @param parameters Key/value dictionary with extra parameters.
     */
    public void setParameters(Map<String, String> parameters) {
		setDirtyIf(!Objects.equals(this.parameters, parameters));
        this.parameters = parameters != null ? Collections.unmodifiableMap(parameters) : null;
    }

    /**
     * Sensor implementation type.
     * @return Sensor implementation type.
     */
    public ProviderSensorRunnerType getType() {
        return type;
    }

    /**
     * Sets the sensor implementation type.
     * @param type Sensor type.
     */
    public void setType(ProviderSensorRunnerType type) {
		this.setDirtyIf(this.type != type);
        this.type = type;
    }

    /**
     * Returns a java class name for a sensor runner.
     * @return Java class for a sensor runner.
     */
    public String getJavaClassName() {
        return javaClassName;
    }

    /**
     * Sets a java class for a sensor runner.
     * @param javaClassName Sensor runner.
     */
    public void setJavaClassName(String javaClassName) {
		this.setDirtyIf(!Objects.equals(this.javaClassName, javaClassName));
        this.javaClassName = javaClassName;
    }

    /**
     * Returns true if the sensor supports grouping by the data stream.
     * @return True when the sensor supports grouping by the data stream.
     */
    public Boolean getSupportsGrouping() {
        return supportsGrouping;
    }

    /**
     * Sets the flag if the sensor supports grouping by the data stream.
     * @param supportsGrouping True when the sensor supports grouping, false otherwise.
     */
    public void setSupportsGrouping(Boolean supportsGrouping) {
        this.setDirtyIf(!Objects.equals(this.supportsGrouping, supportsGrouping));
        this.supportsGrouping = supportsGrouping;
    }

    /**
     * Checks if the sensor supports grouping by the date column identified by the 'partition_by_column' parameter, that means the sensor can analyze daily and monthly partitioned data.
     * @return True when the sensor supports partitioned checks.
     */
    public Boolean getSupportsPartitionedChecks() {
        return supportsPartitionedChecks;
    }

    /**
     * Sets the flag to enable/disable grouping by a partitioning column, enabling support for partitioned checks.
     * @param supportsPartitionedChecks Supports partitioned checks.
     */
    public void setSupportsPartitionedChecks(Boolean supportsPartitionedChecks) {
        this.setDirtyIf(!Objects.equals(this.supportsPartitionedChecks, supportsPartitionedChecks));
        this.supportsPartitionedChecks = supportsPartitionedChecks;
    }

    /**
     * Returns true if this sensor should not be merged with other queries.
     * @return True when this query should not be merged with other queries.
     */
    public boolean isDisableMergingQueries() {
        return disableMergingQueries;
    }

    /**
     * Sets a flag to disable merging queries with other similar queries.
     * @param disableMergingQueries Disable merging queries flag.
     */
    public void setDisableMergingQueries(boolean disableMergingQueries) {
        this.setDirtyIf(this.disableMergingQueries != disableMergingQueries);
        this.disableMergingQueries = disableMergingQueries;
    }

    /**
     * Returns the child map on the spec class with all fields.
     *
     * @return Return the field map.
     */
    @Override
    protected ChildHierarchyNodeFieldMap getChildMap() {
        return FIELDS;
    }

    /**
     * Calls a visitor (using a visitor design pattern) that returns a result.
     *
     * @param visitor   Visitor instance.
     * @param parameter Additional parameter that will be passed back to the visitor.
     */
    @Override
    public <P, R> R visit(HierarchyNodeResultVisitor<P, R> visitor, P parameter) {
        return visitor.accept(this, parameter);
    }

    /**
     * Creates a trimmed version of the object without unwanted properties.
     * A trimmed version is passed to a Jinja2 sql template as a context parameter.
     * @return Trimmed version of this object.
     */
    public ProviderSensorDefinitionSpec trim() {
        return this; // returns self
    }

    /**
     * Creates and returns a deep clone (copy) of this object.
     */
    @Override
    public ProviderSensorDefinitionSpec deepClone() {
        return (ProviderSensorDefinitionSpec)super.deepClone();
    }
}

File: distribution/python/dqops/client/models/provider_sensor_definition_spec.py
from typing import TYPE_CHECKING, Any, Dict, List, Type, TypeVar, Union

from attrs import define as _attrs_define
from attrs import field as _attrs_field

from ..models.provider_sensor_runner_type import ProviderSensorRunnerType
from ..types import UNSET, Unset

if TYPE_CHECKING:
    from ..models.provider_sensor_definition_spec_parameters import (
        ProviderSensorDefinitionSpecParameters,
    )


T = TypeVar("T", bound="ProviderSensorDefinitionSpec")


@_attrs_define
class ProviderSensorDefinitionSpec:
    """
    Attributes:
        type (Union[Unset, ProviderSensorRunnerType]):
        java_class_name (Union[Unset, str]): Java class name for a sensor runner that will execute the sensor. The
            "type" must be "java_class".
        supports_grouping (Union[Unset, bool]): The sensor supports grouping, using the GROUP BY clause in SQL. Sensors
            that support a GROUP BY condition can capture separate data quality scores for each data group. The default
            value is true, because most of the data quality sensor support grouping.
        supports_partitioned_checks (Union[Unset, bool]): The sensor supports grouping by a partition date, using the
            GROUP BY clause in SQL. Sensors that support grouping by a partition_by_column can be used for partition checks,
            calculating separate data quality metrics for each daily/monthly partition. The default value is true, because
            most of the data quality sensor support partitioned checks.
        parameters (Union[Unset, ProviderSensorDefinitionSpecParameters]): Additional provider specific sensor
            parameters
        disable_merging_queries (Union[Unset, bool]): Disables merging this sensor's SQL with other sensors. When this
            parameter is 'true', the sensor's SQL will be executed as an independent query.
    """

    type: Union[Unset, ProviderSensorRunnerType] = UNSET
    java_class_name: Union[Unset, str] = UNSET
    supports_grouping: Union[Unset, bool] = UNSET
    supports_partitioned_checks: Union[Unset, bool] = UNSET
    parameters: Union[Unset, "ProviderSensorDefinitionSpecParameters"] = UNSET
    disable_merging_queries: Union[Unset, bool] = UNSET
    additional_properties: Dict[str, Any] = _attrs_field(init=False, factory=dict)

    def to_dict(self) -> Dict[str, Any]:
        type: Union[Unset, str] = UNSET
        if not isinstance(self.type, Unset):
            type = self.type.value

        java_class_name = self.java_class_name
        supports_grouping = self.supports_grouping
        supports_partitioned_checks = self.supports_partitioned_checks
        parameters: Union[Unset, Dict[str, Any]] = UNSET
        if not isinstance(self.parameters, Unset):
            parameters = self.parameters.to_dict()

        disable_merging_queries = self.disable_merging_queries

        field_dict: Dict[str, Any] = {}
        field_dict.update(self.additional_properties)
        field_dict.update({})
        if type is not UNSET:
            field_dict["type"] = type
        if java_class_name is not UNSET:
            field_dict["java_class_name"] = java_class_name
        if supports_grouping is not UNSET:
            field_dict["supports_grouping"] = supports_grouping
        if supports_partitioned_checks is not UNSET:
            field_dict["supports_partitioned_checks"] = supports_partitioned_checks
        if parameters is not UNSET:
            field_dict["parameters"] = parameters
        if disable_merging_queries is not UNSET:
            field_dict["disable_merging_queries"] = disable_merging_queries

        return field_dict

    @classmethod
    def from_dict(cls: Type[T], src_dict: Dict[str, Any]) -> T:
        from ..models.provider_sensor_definition_spec_parameters import (
            ProviderSensorDefinitionSpecParameters,
        )

        d = src_dict.copy()
        _type = d.pop("type", UNSET)
        type: Union[Unset, ProviderSensorRunnerType]
        if isinstance(_type, Unset):
            type = UNSET
        else:
            type = ProviderSensorRunnerType(_type)

        java_class_name = d.pop("java_class_name", UNSET)

        supports_grouping = d.pop("supports_grouping", UNSET)

        supports_partitioned_checks = d.pop("supports_partitioned_checks", UNSET)

        _parameters = d.pop("parameters", UNSET)
        parameters: Union[Unset, ProviderSensorDefinitionSpecParameters]
        if isinstance(_parameters, Unset):
            parameters = UNSET
        else:
            parameters = ProviderSensorDefinitionSpecParameters.from_dict(_parameters)

        disable_merging_queries = d.pop("disable_merging_queries", UNSET)

        provider_sensor_definition_spec = cls(
            type=type,
            java_class_name=java_class_name,
            supports_grouping=supports_grouping,
            supports_partitioned_checks=supports_partitioned_checks,
            parameters=parameters,
            disable_merging_queries=disable_merging_queries,
        )

        provider_sensor_definition_spec.additional_properties = d
        return provider_sensor_definition_spec

    @property
    def additional_keys(self) -> List[str]:
        return list(self.additional_properties.keys())

    def __getitem__(self, key: str) -> Any:
        return self.additional_properties[key]

    def __setitem__(self, key: str, value: Any) -> None:
        self.additional_properties[key] = value

    def __delitem__(self, key: str) -> None:
        del self.additional_properties[key]

    def __contains__(self, key: str) -> bool:
        return key in self.additional_properties
